framework/logging.cpp

```diff
@@ -1360,20 +1364,21 @@ static void log_data_common(const char *message, const uint8_t *ptr, size_t size
         break;
     }
 
+    spaces = '\n' + std::move(spaces);
     for (size_t i = 0; i < size; ++i) {
         if (current_output_format() != SandstoneApplication::OutputFormat::key_value) {
             if ((i % 32) == 0)
-                fprintf(log, "\n%s", spaces.c_str());
+                buffer += spaces;
             else if ((i % 16) == 0)
-                fputs("  ", log);
+                buffer += "  ";
             else if ((i % 4) == 0)
-                fputs(" ", log);
+                buffer += ' ';
         }
-        fprintf(log, " %02x", (unsigned)ptr[i]);
+        buffer += stdprintf(" %02x", (unsigned)ptr[i]);
     }
 
-    fputc('\n', log);
-    fputc(0, log);
+    buffer += std::string_view("\n", 2);    // include the null
+    fwrite(buffer.data(), 1, msg.size(), log);
```

Comment:
That was `buffer`. I never compiled this commit, because this line went away in another commit of mine.